tests/metrics/test_metrics.py

```diff
@@ -186,3 +193,93 @@ def test_saving_pickable(tmpdir, metric: Metric):
 
     # Check metric value is the same
     assert results_before_save == results_after_load
+
+
+def check_call_order():
+    class DummyMetric(Metric):
+        def __init__(self):
+            super().__init__("dummy")
+            self.call_history = ["init"]
+
+        @staticmethod
+        def input_convert(self, data: Any):
+            self.call_history.append("input_convert")
+            return super(DummyMetric, self).input_convert(self, data)
+
+        def forward(self, tensor1, tensor2):
+            self.call_history.append("forward")
+            return tensor1 - tensor2
+
+        @staticmethod
+        def output_convert(self, data: Any, output: Any):
+            self.call_history.append("output_convert")
+            return super(DummyMetric, self).output_convert(self, data, output)
+
+        def ddp_sync(self, tensor: Any):
+            self.call_history.append("ddp_sync")
+            return super().ddp_sync(tensor)
+
+        @staticmethod
+        def ddp_reduce(self, data: Any, output: Any):
+            self.call_history.append("ddp_reduce")
+            return super(DummyMetric, self).ddp_reduce(self, data, output)
+
+        def aggregate(self, *tensors: torch.Tensor) -> torch.Tensor:
+            self.call_history.append("aggregate")
+            return super().aggregate(*tensors)
+
+        def reset(self):
+            self.call_history.append("reset")
+            return super().reset()
+
+        @property
+        def aggregated(self) -> torch.Tensor:
+            self.call_history.append("aggregated")
+            return super().aggregated
+
+    metric = DummyMetric()
+    assert metric.call_history == ["init"]
+    result = metric(torch.tensor([2.0]), torch.tensor([1.0]))
+    assert torch.allclose(result, torch.tensor(1.0))
+    assert metric.call_history == [
+        "init",
+        "input_convert",
+        "forward",
+        "output_convert",
+        "ddp_reduce",
+        "ddp_sync",
+        "aggregate",
+    ]
+    aggr = metric.aggregated
+    assert metric.call_history == [
+        "init",
+        "input_convert",
+        "forward",
+        "output_convert",
+        "ddp_reduce",
+        "ddp_sync",
+        "aggregate",
+        "aggregated",
+        "aggregate",
+        "reset",
+    ]
+    assert torch.allclose(aggr, result)
+    _ = metric(torch.tensor(2.0), torch.tensor(1.0))
+    assert metric.call_history == [
+        "init",
+        "input_convert",
+        "forward",
+        "output_convert",
+        "ddp_reduce",
+        "ddp_sync",
+        "aggregate",
+        "aggregated",
+        "aggregate",
+        "reset",
+        "input_convert",
+        "forward",
+        "output_convert",
+        "ddp_reduce",
+        "ddp_sync",
+        "aggregate",
+    ]
```

Comment:
Could we add a third test: reinit metric to reset call history, two calls to the metric and then agggregate (case that most corresponds to trying to accumulate over batches

Reply:
sure, will add that one :)

Reply:
done :)